orttraining/orttraining/core/optimizer/megatron_transformer.cc

```diff
@@ -577,23 +733,388 @@ Status MegatronTransformer::TransformSelfAttention(Graph& graph, bool& modified,
+      continue;
+    }
+
+    Node* q_matmul_input_node_ptr = const_cast<Node*>(graph.GetProducerNode(node.MutableInputDefs()[0]->Name()));
+    if (q_matmul_input_node_ptr != nullptr && q_matmul_input_node_ptr->OpType().compare("MegatronF") == 0) {
+      continue;
+    }
+    std::vector<Node*> sub_graph_node_ptrs;
+    sub_graph_node_ptrs.push_back(&node);
+    ProviderType provider_type = node.GetExecutionProviderType();
+
+    std::vector<NodeInfo> linear_pattern = {
+        NodeInfo({add_info}),
+        NodeInfo({mul_info}),
+        NodeInfo({reshape_info}),
+        NodeInfo({transpose_info}),
+        NodeInfo({matmul_info}),
+        NodeInfo({add_info}, false),  // -13
+        NodeInfo({reshape_info}),
+        NodeInfo({where_info}),
+        NodeInfo({reshape_info}),
+        NodeInfo({softmax_info}),
+        NodeInfo({dropout_info}, false),  // -8
+        NodeInfo({matmul_info}),
+        NodeInfo({add_info}, false),
+        NodeInfo({transpose_info}),
+        NodeInfo({reshape_info}),
+        NodeInfo({matmul_info}),  // -3
+        NodeInfo({add_info}),
+        NodeInfo({dropout_info}, false)};  // -1
+    if (!MatchLinearPattern(graph, &node, provider_type, linear_pattern, sub_graph_node_ptrs)) {
+      continue;
+    }
+    // Get all useful nodes here as more vector push back below will change the index.
+    Node* q_biasadd_node_ptr = sub_graph_node_ptrs[sub_graph_node_ptrs.size() - 18];
+    Node& q_transpose_after_reshape_node = *sub_graph_node_ptrs[sub_graph_node_ptrs.size() - 15];
+    Node* qk_matmul_node_ptr = sub_graph_node_ptrs[sub_graph_node_ptrs.size() - 14];
+    Node* dropout_node_ptr = sub_graph_node_ptrs[sub_graph_node_ptrs.size() - 8];
+    Node* qkv_matmul_node_ptr = sub_graph_node_ptrs[sub_graph_node_ptrs.size() - 7];
+    Node& transpose_node1 = *sub_graph_node_ptrs[sub_graph_node_ptrs.size() - 5];
+    Node& dense_matmul_node = *sub_graph_node_ptrs[sub_graph_node_ptrs.size() - 3];
+
+    // Transpose node attribute checking.
+    if (!optimizer_utils::IsAttributeWithExpectedValues(q_transpose_after_reshape_node, "perm", {1LL, 0LL, 2LL}) ||
+        !optimizer_utils::IsAttributeWithExpectedValues(transpose_node1, "perm", {1LL, 0LL, 2LL})) {
+      continue;
+    }
+    // map between reshape node and dim of reshape that must be modified
+    std::unordered_map<Node*, int64_t> reshape_node_ptrs;
+    reshape_node_ptrs[sub_graph_node_ptrs[sub_graph_node_ptrs.size() - 16]] = 1;
+    reshape_node_ptrs[sub_graph_node_ptrs[sub_graph_node_ptrs.size() - 12]] = 1;
+    reshape_node_ptrs[sub_graph_node_ptrs[sub_graph_node_ptrs.size() - 10]] = 0;
+    reshape_node_ptrs[sub_graph_node_ptrs[sub_graph_node_ptrs.size() - 4]] = 2;
+    // till now node should be q matmul operation
+
+    std::vector<Node*> weight_transpose_node_ptrs;
+    std::vector<Node*> bias_add_node_ptrs;
+
+    Node* q_transpose_ptr = const_cast<Node*>(graph.GetProducerNode(node.MutableInputDefs()[1]->Name()));
```

Comment:
we'd better use ORT_ENFORCE non-null check?

Reply:
added additional checks for null as well as compatible op version/domain/provider type.